lib/logsource.c

```diff
@@ -55,15 +59,17 @@ log_source_window_empty(LogSource *self)
 static inline void
 _flow_control_window_size_adjust(LogSource *self, guint32 window_size_increment)
 {
-  guint32 old_window_size;
+  gboolean suspended;
+  gsize old_window_size = window_size_counter_add(&self->window_size, window_size_increment, &suspended);
 
-  window_size_increment += g_atomic_counter_get(&self->suspended_window_size);
-  old_window_size = g_atomic_counter_exchange_and_add(&self->window_size, window_size_increment);
-  g_atomic_counter_set(&self->suspended_window_size, 0);
+  msg_debug("window_size_adjust", evt_tag_int("window_size",
+                                             window_size_counter_get(&self->window_size, NULL)),
+           evt_tag_str("suspended:", suspended ? "TRUE" : "FALSE"));
 
-  if (old_window_size == 0)
+  if (suspended)
     log_source_wakeup(self);
-  if (g_atomic_counter_get(&self->window_size) == self->options->init_window_size)
+
+  if (old_window_size == self->options->init_window_size)
```

Comment:
The new (incremented) window size should be used here instead of the old one.

Instead of calling `window_size_counter_get` here, we could write `old_window_size + window_size_increment == ...` which is guaranteed to preserve the value of the previous `window_size_counter_add` call.

Reply:
yes, you are right